source/draw/gpu/PyramidImage.ooc

```diff
@@ -0,0 +1,166 @@
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use ooc-math
+use ooc-base
+use ooc-collections
+import math
+import RasterMonochrome
+import RasterYuv420Semiplanar
+import gpu/GpuYuv420Semiplanar
+import gpu/GpuMonochrome
+import Image
+
+PyramidImage: class extends Image {
+	_fullRepresentation: RasterMonochrome
+	_validRaster := false
+	fullRepresentation: RasterMonochrome {
+		get {
+			if (!this _validRaster)
+				this _toRaster()
+			this _fullRepresentation
+		}
+	}
+	_fullRepresentationGpu: GpuMonochrome
+	fullRepresentationGpu ::= this _fullRepresentationGpu
+
+	_base: RasterMonochrome
+	base: RasterMonochrome {
+		get {
+			if (this _base == null) {
+				this _base = this _getRepresentationSlice(0, this scaledResolution)
+				this _base referenceCount increase()
+			}
+			this _base
+		}
+	}
+	_representation: RasterMonochrome
+	representation: RasterMonochrome {
+		get {
+			if (this _representation == null) {
+				this _representation = this _getRepresentationSlice(this _scaledResolution area, IntSize2D new(this _scaledResolution width, this _representationRows))
+				this _representation referenceCount increase()
+			}
+			this _representation
+		}
+	}
+
+	//TODO: Not pretty, but I couldn't get this to work with any array type
+	_level1: RasterMonochrome
+	_level2: RasterMonochrome
+	_level3: RasterMonochrome
+	_level4: RasterMonochrome
+	getLevel: func (level: Int) -> RasterMonochrome {
+		result: RasterMonochrome
+		size := this _scaledResolution
+		offset := 0
+		for (i in 0..level) {
+			offset += size area
+			size /= 2
+		}
+		match(level) {
+			case 0 =>
+			 result = this base
+			case 1 =>
+				if (this _level1 == null) {
+					this _level1 = this _getRepresentationSlice(offset, size)
+					this _level1 referenceCount increase()
+				}
+				result = this _level1
+			case 2 =>
+				if (this _level2 == null) {
+					this _level2 = this _getRepresentationSlice(offset, size)
+					this _level2 referenceCount increase()
+				}
+				result = this _level2
+			case 3 =>
+				if (this _level3 == null) {
+					this _level3 = this _getRepresentationSlice(offset, size)
+					this _level3 referenceCount increase()
+				}
+				result = this _level3
+			case 4 =>
+				if (this _level4 == null) {
+					this _level4 = this _getRepresentationSlice(offset, size)
+					this _level4 referenceCount increase()
+				}
+				result = this _level4
```

Comment:
Just for discussions sake, we don't normally do any exception handling or checking of input values. So why should we create default cases?